crates/kubelet/src/store/composite/mod.rs

```diff
@@ -0,0 +1,139 @@
+//! `composite` implements building complex stores from simpler ones.
+
+use crate::store::PullPolicy;
+use crate::store::Store;
+use async_trait::async_trait;
+use oci_distribution::Reference;
+use std::sync::Arc;
+
+/// A `Store` that has additional logic to determine if it can satisfy
+/// a particular reference. An `InterceptingStore` can be composed with
+/// another Store to satisfy specific requests in a custom way.
+pub trait InterceptingStore: Store {
```

Comment:
Is an interceptor model better here or should we just have the separate "file system only" implementation of store? This feels like a lot of code (including 2 publicly exposed traits) for something that might be better as a the separate implementation

Reply:
Just to clarify, there's no "filesystem only" implementation of `Store`.  There's a mode where you can get things from the filesystem _as well as_ registries.  But as written this doesn't turn off registries, because you may have some established modules in a registry and some under-development modules on the filesystem (this was certainly the case with the integration tests while the wasmerciser was under construction).

That aside, it's a good question: rather than offering a mechanism for arbitrary composition, should we just offer two Store implementations, registry-only and hybrid?  Internally the hybrid one would certainly still sit over the existing `FileStore` - no need to reimplement that stuff - and intercept certain requests, but it could be hardwired rather than generic.  And even if internally it was generic, we don't have to surface that composability to kubelet implementers.

I'm kind of agnostic about this.  It sort of irked me to build a one-off composite rather than a general mechanism for composition, which is why I ended up going this route.  But we have no other use case for composition, and as you observe it does increase the public API.

Anyone else have opinions?

Reply:
I am not sure if there are many other needs for a hybrid/composite repo in normal k8s land, but it could be really valuable here. I'm imagining fetching modules from various sources for IoT/Edge devices (even the current FS + registries model is great for that). 

@bacongobbler You have any opinions here?

Reply:
To be honest, OCI registry is not an ideal storage format for WASM modules. And we've seen CapOne and Wasmer already introduce competing module distribution services. So there is a good chance that at some point in the not-too-distant future we are going to need to be able to fetch modules from other network-centric storage systems.

Reply:
Ok, then I think composeability is the right course here @itowlson. Let's keep it as is